### ISSUE TITLE
Bug with renderUI when bookmarking: inputs created dynamically keep reverting back to the bookmarked value

### ISSUE THREAD

**Issue by daattali** [rstudio/shiny#2322]
This one took quite a while to isolate and find the issue :thinking: 

Here's the simplest example I could think of. Using this shiny app:

```
library(shiny)

ui <- function(req) {
  fluidPage(
    bookmarkButton(),
    numericInput("num", "Number", 3),
    uiOutput("rows")
  )
}

server <- function(input, output, session) {
  output$rows <- renderUI({
    lapply(seq(input$num), function(idx) {
      textInput(paste0("text_", idx), paste("Input", idx), "test")
    })
  })
}

shinyApp(ui, server, enableBookmarking = "url")
```

The numeric input controls how many text inputs are shown. Every time you change the number, the text inputs are re-rendered, with the default value of "test". This is fine initially when you run the app **unless the app is restored from a bookmark**.

The problem:

Change any of the text fields, and bookmark the URL. Now load the app using this bookmarked URL. The app will correctly load into the last state. If you change the number however, the text inputs are now going to be initialized using the values that they had when you restored the app from the bookmark, instead of using the values that the server code is telling them to have.

This is most likely happening because of the call to `restoreInput()` inside the text input, which essentially makes the renderUI ignore the value that we pass and just go back to the bookmarked state.

**Comment by daattali** [comment]
Actually this can be reproduced in a much simpler app:

```
library(shiny)

ui <- function(req) {
  fluidPage(
    bookmarkButton(),
    uiOutput("out"),
    actionButton("go", "Go")
  )
}

server <- function(input, output, session) {
  output$out <- renderUI({
    input$go
    textInput("text", "Text", "text")
  })
}

shinyApp(ui, server, enableBookmarking = "url")
```

Pressing the button should re-render the input with "text". After restoring from a bookmark, it will instead always render the input with the bookmarked value.

![image](https://user-images.githubusercontent.com/952340/52391652-5a388980-2a6c-11e9-8be7-36b344165aad.png)

**Comment by alandipert** [comment]
Thanks @daattali, I reproduced. I strongly suspect this was introduced by 0bb53e8ca5b8e769b11a0ec37a5b5c66eceb9a54

**Comment by wch** [comment]
Fixed in the rc-v1.3.0 branch by #2360.
